### PR TITLE
Add tentative support for OCaml 5.3

### PR DESCRIPTION
**Pull request by cristianoc**
The cmt format has changed, this tries to support value dependencies just based on the type of what's available in the new `Cmt_format.cmt_infos`.

https://github.com/rescript-lang/reanalyze/issues/202

### CODE REVIEW

File: src/Compat.ml [hunk 1]
```diff
@@ -241,3 +262,29 @@ let get_desc = Types.get_desc
...
+let constant_desc d =
+#if OCAML_VERSION >= (5, 3, 0)
+  d.Parsetree.pconst_desc
+#else
+  d
+#endif
+
+let extractValueDependencies (cmt_infos : CL.Cmt_format.cmt_infos) =
+#if OCAML_VERSION >= (5, 3, 0)
+  let deps = ref [] in
```
Comment (cristianoc) [inline src/Compat.ml]: This tentatively tries to extract the old `cmt_value_dependencies`.

**Comment by cristianoc** [comment]
The dependencies seems to have changed, currently none are found, and this attempt does not seem to work either:

```ml
let extractValueDependencies (cmt_infos : Cmt_format.cmt_infos) =
#if OCAML_VERSION >= (5, 3, 0)
  let deps = ref [] in
  let ident_occurrences = cmt_infos.cmt_ident_occurrences in
  Printf.eprintf "ident_occurrences: %d\n" (List.length ident_occurrences);
  let process_id ({Location.txt; loc}, res) = match res with
     | Shape_reduce.Resolved uid | Resolved_alias (uid, _) | Approximated (Some uid) ->
      (match Types.Uid.Tbl.find_opt cmt_infos.cmt_uid_to_decl uid with  
      | Some (Value v) ->
        deps := (loc, v.val_loc) :: !deps
      | Some (Type td) ->
        deps := (loc, td.typ_loc) :: !deps
      | Some (Value_binding vb) ->
        deps := (loc, vb.vb_loc) :: !deps
      | Some (Module_binding mb) ->
        deps := (loc, mb.mb_loc) :: !deps
      | Some (Module_type mt) ->
        deps := (loc, mt.mtd_loc) :: !deps
      | Some (Constructor cd) ->
        deps := (loc, cd.cd_loc) :: !deps
      | Some (Class c) ->
        deps := (loc, c.ci_loc) :: !deps
      | Some (Class_type ct) ->
        deps := (loc, ct.ci_loc) :: !deps
      | Some (Extension_constructor ec) ->
        deps := (loc, ec.ext_loc) :: !deps
      | Some (Label ld) ->
        deps := (loc, ld.ld_loc) :: !deps
      | Some (Module m) ->
        deps := (loc, m.md_loc) :: !deps
      | Some (Module_substitution ms) ->
        deps := (loc, ms.ms_loc) :: !deps
      | _ -> ())
     | _ -> () in
  List.iter process_id ident_occurrences;
  List.rev !deps
```

Moving this to draft.

**Comment by fantazio** [comment]
Hi @cristianoc ! Shameless plug here, this problem has been solved in the dead_code_analyzer (https://github.com/LexiFi/dead_code_analyzer/pull/38) and you might be inspired by a similar solution.

TL;DR: the `cmt_infos.cmt_value_dependencies` is replaced by `cmt_infos.cmt_declaration_dependencies`, and one needs to read the `cmt_infos.cmt_uid_to_decl` to recreate the old association. A very important trick is that the table in the current `cmt_infos` may not be complete (see https://github.com/ocaml/ocaml/blob/5.3.0/typing/shape.mli), in which case it must be read in multiple `.cmt` and `.cmti` files to gather all the necessary information.

**Comment by cristianoc** [comment]
Sweet, thank you!
This seems to work.

**Comment by JonoPrest** [comment]
Hey @cristianoc

I can see this is WIP but I took this branch for a spin on the rescript-compiler itself. I built it against a 5.3 switch, ran dune build @check, then -dce-cmt _build/default. It runs cleanly end to end, but the cross-module reference tracking looks pretty off. It flags a lot of clearly-live stuff as dead. The one that jumped out was ast_helper is a dead module for example. Same thing for Lam_compile and Js_dump. Anything local to a module seems right, but as soon as something is exported via its .mli and used elsewhere, those uses seem to disappear and it gets called dead.

Either way still lots of useable warnings for dce, excited for this to work 💪🏼

**Comment by cristianoc** [comment]
> Hey @cristianoc
> 
> I can see this is WIP but I took this branch for a spin on the rescript-compiler itself. I built it against a 5.3 switch, ran dune build @check, then -dce-cmt _build/default. It runs cleanly end to end, but the cross-module reference tracking looks pretty off. It flags a lot of clearly-live stuff as dead. The one that jumped out was ast_helper is a dead module for example. Same thing for Lam_compile and Js_dump. Anything local to a module seems right, but as soon as something is exported via its .mli and used elsewhere, those uses seem to disappear and it gets called dead.
> 
> Either way still lots of useable warnings for dce, excited for this to work 💪🏼

Might be worth checking whether it has to do with this (copied from the readme):
> Make sure that dune builds both .cmt and .cmti files by enabling bytecode compilation. This is normally done by adding (modes byte exe) to the executable stanza in your dune file (see https://github.com/ocaml/dune/issues/3182):

**Comment by JonoPrest** [comment]
> Might be worth checking whether it has to do with this (copied from the readme):
> 
> > Make sure that dune builds both .cmt and .cmti files by enabling bytecode compilation. This is normally done by adding (modes byte exe) to the executable stanza in your dune file (see [ocaml/dune#3182](https://github.com/ocaml/dune/issues/3182)):

I ran it with `dune build @check`, and I can see all the .cmt and .cmi files in `_build/default`

**Comment by cknitt** [comment]
@cristianoc What is still open here? Could it be merged and released already?

**Comment by cristianoc** [comment]
> @cristianoc What is still open here? Could it be merged and released already?

Should be ready  to go. I'll merge now.
Can't remember though what's required to trigger a new release so opam finds it.
